bee-protocol/src/worker/peer/message_handler.rs

```diff
@@ -126,31 +130,38 @@ impl EventHandler {
     }
     /// Fetch a slice of bytes of a determined length.
     ///
-    /// This method returns `None` if a shutdown signal was received.
-    async fn fetch_bytes<'a>(&'a mut self, len: usize) -> Option<&'a [u8]> {
+    /// The future returned by this method will be ready until there are enough bytes to fullfill
+    /// the request.
+    async fn fetch_bytes<'a>(&'a mut self, len: usize) -> &'a [u8] {
         // We need to be sure that we have enough bytes in the buffer.
         while self.offset + len > self.buffer.len() {
-            // If there are not enough bytes in the buffer, we must either receive new events until
-            // we have enough bytes, or receive a shutdown signal.
-            select! {
-                event = self.receiver.next() => {
-                    // If we received an event, we push it to the buffer.
-                    if let Some(event) = event {
-                        self.push_event(event);
-                    }
-                },
-                _ = &mut self.shutdown => {
-                    // If we received a shutdown signal, we return `None`.
-                    return None;
-                }
+            // If there are not enough bytes in the buffer, we must receive new events
+            if let Some(event) = self.receiver.next().await {
+                // If we received an event, we push it to the buffer.
+                self.push_event(event);
             }
         }
         // Get the requested bytes. This will not panic because the loop above only exists if we
         // have enough bytes to do this step.
         let bytes = &self.buffer[self.offset..(self.offset + len)];
         // Increase the offset by the length of the byte slice.
         self.offset += len;
-        Some(bytes)
+        bytes
+    }
+
+    /// Helper method to be able to shutdown when fetching bytes for a message.
+    ///
+    /// This method returns `None` if a shutdown signal is received, otherwise it returns the
+    /// requested bytes.
+    async fn fetch_bytes_or_shutdown<'a>(
+        &'a mut self,
+        mut shutdown: &'a mut ShutdownRecv,
+        len: usize,
+    ) -> Option<&'a [u8]> {
+        select! {
+            _ = shutdown => None,
+            bytes = self.fetch_bytes(len).fuse() => Some(bytes),
```

Comment:
From https://docs.rs/futures-preview/0.3.0-alpha.19/futures/macro.select.html
> Futures and streams which are not already fused can be fused using the .fuse() method. Note, though, that fusing a future or stream directly in the call to select! will not be enough to prevent it from being polled after completion if the select! call is in a loop, so when select!ing in a loop, users should take care to fuse() outside of the loop.

Is this a concern ?

Reply:
to be honest, I'm not sure.

Reply:
I added a comment to be sure we never flip the order of the select.